### PR TITLE
Implement Serializers for X25519 and X448

### PR DESCRIPTION
**Pull request by mattcaswell**
Provide serializers for X25519 and X448 for text, pem and der. There are
no paramater serializers because there are no parameters for these
algorithms.

### CODE REVIEW

File: providers/implementations/serializers/serializer_ecx.c [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    if (type == ECX_KEY_TYPE_X25519) {
+        *ecx_new = ossl_prov_get_keymgmt_new(x25519_keymgmt_functions);
+        *ecx_free = ossl_prov_get_keymgmt_free(x25519_keymgmt_functions);
+        *ecx_import = ossl_prov_get_keymgmt_import(x25519_keymgmt_functions);
+    } else {
+        *ecx_new = ossl_prov_get_keymgmt_new(x448_keymgmt_functions);
+        *ecx_free = ossl_prov_get_keymgmt_free(x448_keymgmt_functions);
+        *ecx_import = ossl_prov_get_keymgmt_import(x448_keymgmt_functions);
+    }
+}
```
Comment (levitte) [inline providers/implementations/serializers/serializer_ecx.c]: Why didn't I think of this???

File: providers/implementations/serializers/serializer_local.h [hunk 2]
```diff
@@ -49,6 +55,13 @@ int ossl_prov_prepare_dh_params(const void *dh, int nid,
...
 
+void ecx_get_new_free_import(ECX_KEY_TYPE type,
+                             OSSL_OP_keymgmt_new_fn **ecx_new,
+                             OSSL_OP_keymgmt_free_fn **ecx_free,
+                             OSSL_OP_keymgmt_import_fn **ecx_import);
+int ossl_prov_ecx_pub_to_der(const void *ecxkey, unsigned char **pder);
+int ossl_prov_ecx_priv_to_der(const void *ecxkey, unsigned char **pder);
+
 int ossl_prov_prepare_dsa_params(const void *dsa, int nid,
                                 ASN1_STRING **pstr, int *pstrtype);
```
Comment (levitte) [inline providers/implementations/serializers/serializer_local.h]: Nit: indent one space off

**Comment by slontis** [comment]
looks like no-ec is broken in travis

File: providers/implementations/serializers/serializer_local.h (hunk 2 above)
Comment (mattcaswell) [reply providers/implementations/serializers/serializer_local.h]: This is existing code...but ok

**Comment by mattcaswell** [comment]
Fixup pushed to address the feedback and to correct the no-ec build.

**Comment by mspncp** [comment]
Typo in commit message: paramater -> parameter

**Comment by mattcaswell** [comment]
Squash commit pushed fixing the typo

File: providers/implementations/serializers/serializer_local.h (hunk 2 above)
Comment (levitte) [reply providers/implementations/serializers/serializer_local.h]: Oops. Sorry

File: providers/implementations/serializers/serializer_ecx.c [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+void ecx_get_new_free_import(ECX_KEY_TYPE type,
+                             OSSL_OP_keymgmt_new_fn **ecx_new,
+                             OSSL_OP_keymgmt_free_fn **ecx_free,
+                             OSSL_OP_keymgmt_import_fn **ecx_import)
+{
+    if (type == ECX_KEY_TYPE_X25519) {
+        *ecx_new = ossl_prov_get_keymgmt_new(x25519_keymgmt_functions);
+        *ecx_free = ossl_prov_get_keymgmt_free(x25519_keymgmt_functions);
+        *ecx_import = ossl_prov_get_keymgmt_import(x25519_keymgmt_functions);
+    } else {
```
Comment (paulidale) [inline providers/implementations/serializers/serializer_ecx.c]: Ideally, a `type == ECX_KEY_TYPE_X448` check here would be good.
There is no way to return a failure, so it's kind of a moot point.

File: providers/implementations/serializers/serializer_ecx_priv.c [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+    return settables;
+}
+
+static int ecx_priv_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    struct ecx_priv_ctx_st *ctx = vctx;
+    const OSSL_PARAM *p;
+
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_SERIALIZER_PARAM_CIPHER))
```
Comment (paulidale) [inline providers/implementations/serializers/serializer_ecx_priv.c]: Breaking this into `p = ...;` and `if (p != NULL)` would add clarity and not increase the line count.

File: providers/implementations/serializers/serializer_ecx.c (hunk 3 above)
Comment (levitte) [reply providers/implementations/serializers/serializer_ecx.c]: Not quite mot... Not setting the pointers would become an assertion later on. This shouldn't ever happen, but if the algo name strings are somehow wrong, we will notice!

Comment (mattcaswell) [reply providers/implementations/serializers/serializer_ecx.c]: Fixed.

File: providers/implementations/serializers/serializer_ecx_priv.c (hunk 4 above)
Comment (mattcaswell) [reply providers/implementations/serializers/serializer_ecx_priv.c]: Fixed

**Comment by mattcaswell** [comment]
Update pushed addressing feedback so far.

File: providers/implementations/serializers/libimplementations-lib-serializer_ecx.d.tmp [hunk 5]
```diff
@@ -0,0 +1,25 @@
+providers/implementations/serializers/libimplementations-lib-serializer_ecx.o: \
```
Comment (paulidale) [inline providers/implementations/serializers/libimplementations-lib-serializer_ecx.d.tmp]: oops

File: providers/implementations/serializers/serializer_ecx_priv.c [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+static int ecx_priv_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    struct ecx_priv_ctx_st *ctx = vctx;
+    const OSSL_PARAM *p;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SERIALIZER_PARAM_CIPHER);
+    if (p != NULL) {
+        const OSSL_PARAM *propsp =
+            OSSL_PARAM_locate_const(params, OSSL_SERIALIZER_PARAM_PROPERTIES);
+        const char *props = NULL;
```
Comment (paulidale) [inline providers/implementations/serializers/serializer_ecx_priv.c]: This variable is assigned NULL by default six lines down.  No need for both.

File: providers/implementations/serializers/serializer_ecx_pub.c [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+/*
+ * Context used for public key serialization.
+ */
+struct ecx_pub_ctx_st {
+    void *provctx;
+    ECX_KEY_TYPE type;
+};
+
+
```
Comment (paulidale) [inline providers/implementations/serializers/serializer_ecx_pub.c]: Spurious blank line.

File: providers/implementations/serializers/libimplementations-lib-serializer_ecx.d.tmp (hunk 5 above)
Comment (levitte) [reply providers/implementations/serializers/libimplementations-lib-serializer_ecx.d.tmp]: this is a good reason not to build in the source directory...  but also to fix .gitignore

**Comment by richsalz** [comment]
> this is a good reason not to build in the source directory.

Woah!  That's not going to become a real thing, is it?

**Comment by levitte** [comment]
> Woah! That's not going to become a real thing, is it?

Don't panic, it's of course still possible to build in the source directory.

File: providers/implementations/serializers/libimplementations-lib-serializer_ecx.d.tmp (hunk 5 above)
Comment (mattcaswell) [reply providers/implementations/serializers/libimplementations-lib-serializer_ecx.d.tmp]: Fixed. Also fixed .gitignore in #11122

File: providers/implementations/serializers/serializer_ecx_priv.c (hunk 6 above)
Comment (mattcaswell) [reply providers/implementations/serializers/serializer_ecx_priv.c]: Fxied

File: providers/implementations/serializers/serializer_ecx_pub.c (hunk 7 above)
Comment (mattcaswell) [reply providers/implementations/serializers/serializer_ecx_pub.c]: Fixed

**Comment by mattcaswell** [comment]
Fixup pushed address above feedback.

**Comment by kroeckx** [comment]
The commit message indicates this is for "text, pem and der", but I see none of them being tested. (I have no idea what text means here.)

Does follow the format from some RFC like RFC8410?

**Comment by levitte** [comment]
Test is printouts, basically. It corresponds to the ameth print methods.

**Comment by kroeckx** [comment]
This seems to lack documentation updates. We seem to have an
EVP_PKEY_fromdata() function, but it's not documented what that
function expects as input.

A function like EVP_PKEY_param_fromdata_settable() is making
little sense to me. That looks like something that needs to be
in the documentation, rather that first writing a test program to
see what parameters I can set.

That whole manpage needs examples.

**Comment by levitte** [comment]
Good point, @kroeckx, thanks for pointing that out.

**Comment by slontis** [comment]
@paulidale does this need reapproval?

**Comment by mattcaswell** [comment]
> @paulidale does this need reapproval?

I need to find some time to look at this again in light of @kroeckx's comments.

**Comment by mattcaswell** [comment]
I've rebased this to resolve a conflict with master and pushed some significant updates. I've expanded out the testing significantly. Previously we tested things by checking we were able to print out private keys in text and pem, as well as public keys in pem. We now generate both public and private keys in text, pem and der and compare them to a "known good" file. I've done this for all the existing serializers (not just X25519 and X448). This identified a few bugs which I've also fixed.

Additionally I've expanded out the documentation for all the various key types to document the import/export types that are available for built-in algorithms.

I've not touched EVP_PKEY_fromdata() as mentioned by @kroeckx. That's not really directly touched by this PR so probably adding examples should be covered by a subsequent PR. I see @levitte has opened an issue for this (#11131).

Note that the "known good" files referred to above are self-generated. I intend to double check these against the 1.1.1 implementation (which doesn't have all this new serializer code) to confirm that they really are good - but I've not yet done that.

Hopefully this addresses the comments that were raised. Please take another look.

**Comment by slontis** [comment]
Travis NITs

File: providers/defltprov.c [hunk 8]
```diff
@@ -464,6 +464,34 @@ static const OSSL_ALGORITHM deflt_serializer[] = {
       dsa_param_pem_serializer_functions },
 #endif
 
+#ifndef OPENSSL_NO_EC
+    { "X25519", "default=yes,format=text,type=private",
```
Comment (slontis) [inline providers/defltprov.c]: these need to change from default= ?

**Comment by kroeckx** [comment]
I'm fine with examples to EVP_PKEY_fromdata being added in an other PR, but I think this PR should document the format EVP_PKEY_fromdata() is expecting.

**Comment by mattcaswell** [comment]
> I'm fine with examples to EVP_PKEY_fromdata being added in an other PR, but I think this PR should document the format EVP_PKEY_fromdata() is expecting.

In response to your comment on this previously I added a description of all the formats to provider-keymgmt.pod (both for the serializers added by this PR, as well as all the existing serlializers). The formats are the same whether you are a provider author wanting to implement serializers that conform to those formats, or whether your are an end application author wanting to use EVP_PKEY_fromdata(). Its probably not a good idea to repeat the documentation in both places. My assumption was that the PR that updates the EVP_PKEY_fromdata() page would include a reference to the provider-keymgmt.pod page.

File: providers/defltprov.c (hunk 8 above)
Comment (mattcaswell) [reply providers/defltprov.c]: Fixed

**Comment by mattcaswell** [comment]
I've rebased this and (hopefully) addressed the travis nits.

> My assumption was that the PR that updates the EVP_PKEY_fromdata() page would include a reference to the provider-keymgmt.pod page.

I've now added that reference as part of this PR.

File: doc/man7/provider-keymgmt.pod [hunk 9]
```diff
@@ -256,6 +256,87 @@ OP_keymgmt_export_types() should return a constant array of descriptor
...
+The following Import/Export types are available for the built-in X25519, X448,
+ED25519 and X448 algorithms:
+
+=over 4
+
+=item "pub" (B<OSSL_PKEY_PARAM_PUB_KEY>) <octet string>
+
+The public key value.
+
+=item "priv" (B<OSSL_PKEY_PARAM_PRIV_KEY>) <integer> or <octet string>
```
Comment (kroeckx) [inline doc/man7/provider-keymgmt.pod]: Does integer make any sense here?

Comment (mattcaswell) [reply doc/man7/provider-keymgmt.pod]: Oops. No. This is a left over from an earlier iteration where I had the description of this combined with DSA/DH version which does use an integer. Will remove.

Comment (mattcaswell) [reply doc/man7/provider-keymgmt.pod]: Fixed.

**Comment by mattcaswell** [comment]
I pushed a fixup addressing the one item of feedback above.

The appveyor failure looks relevant. Unfortunately I do not have access to a Windows environment right at the moment, so I won't be able to investigate it until I get back from the RSA conference. :-(

File: providers/implementations/serializers/serializer_ecx_priv.c [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    struct ecx_priv_ctx_st *ctx = vctx;
+    ECX_KEY *ecxkey = vecxkey;
+    int ret;
+
+    ctx->sc.cb = cb;
+    ctx->sc.cbarg = cbarg;
+
+    ret = ossl_prov_write_priv_der_from_obj(out, ecxkey,
+                                            ctx->type == ECX_KEY_TYPE_X25519
+                                            ? EVP_PKEY_X25519 : EVP_PKEY_X448,
```
Comment (levitte) [inline providers/implementations/serializers/serializer_ecx_priv.c]: It might be worth having a local variable to get that value

**Comment by slontis** [comment]
> The appveyor failure looks relevant. Unfortunately I do not have access to a Windows environment right at the moment, so I won't be able to investigate it until I get back from the RSA conference. :-(

Looks weird..
It only happens with the first key type during X25519.pub.der
It looks like the last 10 bytes of the of bitstring are missing.. (almost looks like the length has been truncated for some reason).

**Comment by mattcaswell** [comment]
@levitte figured it out. It's because I open the file with "r" and not "rb" and so it interprets a "1a" byte as EOF.

**Comment by mattcaswell** [comment]
Fixups pushed to attempt to address the appveyor issue.

File: providers/implementations/serializers/serializer_ecx_priv.c (hunk 10 above)
Comment (mattcaswell) [reply providers/implementations/serializers/serializer_ecx_priv.c]: Fixed

File: providers/implementations/serializers/serializer_common.c [hunk 11]
```diff
@@ -243,6 +243,36 @@ int ossl_prov_print_labeled_bignum(BIO *out, const char *label,
     return 1;
 }
 
+/* Number of octets per line */
+#define LABELED_BUF_PRINT_WIDTH    15
```
Comment (slontis) [inline providers/implementations/serializers/serializer_common.c]: Why do we use such a weird number, why not 16??

Comment (mattcaswell) [reply providers/implementations/serializers/serializer_common.c]: To be consistent with the way `ASN1_buf_print` currently does this.

**Comment by slontis** [comment]
You have a green build on one of the failing windows machines from before - so it looks like it has been addressed.

**Comment by slontis** [comment]
Merged to master.

**Comment by mattcaswell** [comment]
> Merged to master.

Thanks @slontis!

**Pull request closed without merging**